packages/x-data-grid/src/hooks/features/rowSelection/useGridRowSelection.ts

```diff
@@ -214,24 +214,45 @@ export const useGridRowSelection = (
       if (resetSelection) {
         logger.debug(`Setting selection for row ${id}`);
 
-        apiRef.current.setRowSelectionModel(isSelected ? [id] : []);
+        const newSelection: GridRowId[] = [];
+        if (isSelected) {
+          newSelection.push(id);
+          if (props.propagateRowSelection) {
+            const rowsToSelect = findRowsToSelect(apiRef, tree, id);
+            rowsToSelect.forEach((rowId) => {
+              newSelection.push(rowId);
+            });
+          }
+        }
+
+        apiRef.current.setRowSelectionModel(newSelection);
       } else {
         logger.debug(`Toggling selection for row ${id}`);
 
         const selection = gridRowSelectionStateSelector(apiRef.current.state);
-        const newSelection: GridRowId[] = selection.filter((el) => el !== id);
+
+        const newSelection: Set<GridRowId> = new Set(selection.filter((el) => el !== id));
```

Comment:
nice use of `Set` 👍🏼

Reply:
Yes, I found it a performance bottleneck during my performance testing.
The `.filter` here previously was causing de-selection to hang the Data Grid completely. 😅

Reply:
Btw perf wise this line can be simplified to:
```js
const newSelection = new Set(selection)
newSelection.delete(id)
```

Reply:
Nice one 👌
Thanks!